### PR TITLE
[feature-wip](array-type) add function array_remove

### PR DESCRIPTION
**Pull request by xy720**
# Proposed changes

Usage example:

array_remove remove all elements in array which is equal to the target.

```
mysql> set enable_vectorized_engine=true;
```

insert some data into array table

```
mysql> select k1, k2 from array_type_table_1;
+------+--------------------+
| k1   | k2                 |
+------+--------------------+
|    1 | [1, 2, 3]          |
|    2 | [1, 3]             |
|    3 | NULL               |
|    4 | [1, 3]             |
|    5 | [NULL, 1, NULL, 2] |
+------+--------------------+

mysql> select k1, k2, array_remove(k2, 1) from array_type_table_1;
+------+--------------------+-----------------------+
| k1   | k2                 | array_remove(`k2`, 1) |
+------+--------------------+-----------------------+
|    1 | [1, 2, 3]          | [2, 3]                |
|    2 | [1, 3]             | [3]                   |
|    3 | NULL               | NULL                  |
|    4 | [1, 3]             | [3]                   |
|    5 | [NULL, 1, NULL, 2] | [NULL, NULL, 2]       |
+------+--------------------+-----------------------+

mysql> select k1, k2, array_remove(k2, k1) from array_type_table_1;
+------+--------------------+--------------------------+
| k1   | k2                 | array_remove(`k2`, `k1`) |
+------+--------------------+--------------------------+
|    1 | [1, 2, 3]          | [2, 3]                   |
|    2 | [1, 3]             | [1, 3]                   |
|    3 | NULL               | NULL                     |
|    4 | [1, 3]             | [1, 3]                   |
|    5 | [NULL, 1, NULL, 2] | [NULL, 1, NULL, 2]       |
+------+--------------------+--------------------------+
5 rows in set (0.01 sec)

mysql> select k1, k2, array_remove(k2, NULL) from array_type_table_1;
+------+--------------------+--------------------------+
| k1   | k2                 | array_remove(`k2`, NULL) |
+------+--------------------+--------------------------+
|    1 | [1, 2, 3]          | NULL                     |
|    2 | [1, 3]             | NULL                     |
|    3 | NULL               | NULL                     |
|    4 | [1, 3]             | NULL                     |
|    5 | [NULL, 1, NULL, 2] | NULL                     |
+------+--------------------+--------------------------+
5 rows in set (0.01 sec)
```

other usage:

```
mysql> select array_remove(['test', NULL, 'value'], 'value');
+-----------------------------------------------------+
| array_remove(ARRAY('test', NULL, 'value'), 'value') |
+-----------------------------------------------------+
| [test, NULL]                                        |
+-----------------------------------------------------+
1 row in set (0.01 sec)

mysql> select k1, k2, array_remove(k2, date('2022-10-10')) from array_type_table_date;
+------+--------------------------+-------------------------------------------------+
| k1   | k2                       | array_remove(`k2`, date('2022-10-10 00:00:00')) |
+------+--------------------------+-------------------------------------------------+
|    1 | [2021-10-10, 2022-10-10] | [2021-10-10]                                    |
|    2 | [NULL, 2022-05-14]       | [NULL, 2022-05-14]                              |
+------+--------------------------+-------------------------------------------------+
```

related to #10052

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/array/function_array_remove.h [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+//        if (col_left.column->is_nullable()) {
+//            auto null_col = check_and_get_column<ColumnNullable>(*col_left.column);
+//            src_null_map = null_col->get_null_map_column().get_data().data();
+//            args = {{null_col->get_nested_column_ptr(), remove_nullable(col_left.type),
+//                     col_left.name},
+//                    block.get_by_position(arguments[1])};
+//        } else {
+//            args = {col_left, block.get_by_position(arguments[1])};
+//        }
+//        auto res_column = _execute_non_nullable(args, input_rows_count, src_null_map);
```
Comment (adonis0147) [inline be/src/vec/functions/array/function_array_remove.h:68]: Useless code should be removed.

File: be/src/vec/functions/array/function_array_remove.h [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+
+        // For default implementation of nulls args
+        ColumnsWithTypeAndName args = {
+                block.get_by_position(arguments[0]), block.get_by_position(arguments[1])
+        };
+
+        LOG(INFO) << "left type is " << args[0].type->get_name();
+        LOG(INFO) << "right type is " << args[1].type->get_name();
+        LOG(INFO) << "right column type is " << args[1].column->get_name();
```
Comment (adonis0147) [inline be/src/vec/functions/array/function_array_remove.h:77]: Logging in IO path should be simplified.

File: be/src/vec/functions/array/function_array_remove.h (hunk 1 above)
Comment (xy720) [reply be/src/vec/functions/array/function_array_remove.h]: done

File: be/src/vec/functions/array/function_array_remove.h (hunk 2 above)
Comment (xy720) [reply be/src/vec/functions/array/function_array_remove.h]: done

File: be/src/vec/utils/util.hpp [hunk 3]
```diff
@@ -38,6 +40,12 @@ class VectorizedUtils {
         ColumnsWithTypeAndName columns_with_type_and_name;
         for (const auto& tuple_desc : row_desc.tuple_descriptors()) {
             for (const auto& slot_desc : tuple_desc->slots()) {
+                LOG(INFO) << "slot type is " << slot_desc->get_data_type_ptr()->get_name();
```
Comment (carlvinhust2012) [inline be/src/vec/utils/util.hpp:43]: is this log necessary？

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_remove.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+-->
+
+## array_remove
+
+### description
+
+#### Syntax
+
+```
+ARRAY<T> array_remove(ARRAY<T> arr, T)
```
Comment (cambyzju) [inline docs/en/docs/sql-manual/sql-functions/array-functions/array_remove.md:32]:
```suggestion
ARRAY<T> array_remove(ARRAY<T> arr, T val)
```

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_remove.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+```
+
+Remove all elements that equal to element from array.
```
Comment (cambyzju) [inline docs/en/docs/sql-manual/sql-functions/array-functions/array_remove.md:35]: Add more cases, like: If arr is NULL or val is NULL, how array_remove works?

File: be/src/vec/functions/array/function_array_remove.h [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+        // check array nested column type and get data
+        const auto& src_offs = reinterpret_cast<const ColumnString&>(nested_column).get_offsets();
+        const auto& src_chars = reinterpret_cast<const ColumnString&>(nested_column).get_chars();
+
+        // check right column type and get data
+        const auto& target_offs = reinterpret_cast<const ColumnString&>(target).get_offsets();
+        const auto& target_chars = reinterpret_cast<const ColumnString&>(target).get_chars();
+
+        // prepare dst array
+        auto dst = ColumnArray::create(
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_remove.h:137]: need to consider nested column nullable or not.

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_remove.md (hunk 5 above)
Comment (xy720) [reply docs/en/docs/sql-manual/sql-functions/array-functions/array_remove.md:35]: done

File: be/src/vec/functions/array/function_array_remove.h (hunk 6 above)
Comment (xy720) [reply be/src/vec/functions/array/function_array_remove.h]: Ok. I have refactored the code.

File: be/src/vec/functions/array/function_array_remove.h [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        } else if (check_column<ColumnInt128>(right_column)) {
+            return _execute_number<NestedColumnType, ColumnInt128>(offsets, nested_column,
+                                                                   right_column, nested_null_map);
+        } else if (check_column<ColumnFloat32>(right_column)) {
+            return _execute_number<NestedColumnType, ColumnFloat32>(offsets, nested_column,
+                                                                    right_column, nested_null_map);
+        } else if (check_column<ColumnFloat64>(right_column)) {
+            return _execute_number<NestedColumnType, ColumnFloat64>(offsets, nested_column,
+                                                                    right_column, nested_null_map);
+        } else if (right_column.is_date_type()) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_remove.h:259]: put this line before check_column<ColumnInt64>

File: be/src/vec/functions/array/function_array_remove.h [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    ColumnPtr _execute_number(const ColumnArray::Offsets& offsets, const IColumn& nested_column,
+                              const IColumn& right_column, const UInt8* nested_null_map) {
+        // check array nested column type and get data
+        const auto& src_data = reinterpret_cast<const NestedColumnType&>(nested_column).get_data();
+
+        // check target column type and get data
+        const auto& target_data = reinterpret_cast<const RightColumnType&>(right_column).get_data();
+
+        PaddedPODArray<UInt8>* dst_null_map = nullptr;
+        ColumnPtr array_nested_column = nullptr;
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_remove.h:79]:
```suggestion
        MutableColumnPtr array_nested_column = nullptr;
```

File: be/src/vec/functions/array/function_array_remove.h [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+        IColumn* dst_column;
+        if (nested_null_map) {
+            auto dst_nested_column = ColumnNullable::create(nested_column.clone_empty(), ColumnUInt8::create());
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_remove.h:82]:
```suggestion
            array_nested_column = ColumnNullable::create(nested_column.clone_empty(), ColumnUInt8::create());
```

File: be/src/vec/functions/array/function_array_remove.h [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+                    // case: array:[Null], target:1 ==> [Null]
+                    dst_data.push_back(typename NestedColumnType::value_type());
+                    dst_null_map->push_back(1);
+                    continue;
+                }
+
+                if (src_data[off + pos] == target_data[row]) {
+                    ++count;
+                } else {
+                    if (nested_null_map) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_remove.h:123]:
```suggestion
                    if (nested_null_map) {
                        dst_null_map->push_back(0);
                    } 
                    dst_data.push_back(src_data[off + pos]);
                  
```

File: be/src/vec/functions/array/function_array_remove.h [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+            dst_null_map->reserve(offsets.back());
+        } else {
+            auto dst_nested_column = nested_column.clone_empty();
+            array_nested_column = dst_nested_column->get_ptr();
+            dst_column = dst_nested_column;
+        }
+
+        auto& dst_offs = reinterpret_cast<ColumnString&>(*dst_column).get_offsets();
+        auto& dst_chars = reinterpret_cast<ColumnString&>(*dst_column).get_chars();
+        dst_offs.reserve(src_offs.back());
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_remove.h:167]:
```suggestion
        dst_offs.reserve(src_offs.size());
```

File: be/src/vec/functions/array/function_array_remove.h [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+            }
+
+            size_t target_off = target_offs[row - 1];
+            size_t target_len = target_offs[row] - target_off;
+
+            size_t count = 0;
+            for (size_t pos = 0; pos < len; ++pos) {
+                if (nested_null_map && nested_null_map[off + pos]) {
+                    // case: array:[Null], target:'str' ==> [Null]
+                    dst_chars.push_back(0);
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_remove.h:192]: no need to push_back(0)

File: be/src/vec/functions/array/function_array_remove.h [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+
+                size_t src_pos = src_offs[pos + off - 1];
+                size_t src_len = src_offs[pos + off] - src_pos;
+                const char* src_raw_v = reinterpret_cast<const char*>(&src_chars[src_pos]);
+                const char* target_raw_v = reinterpret_cast<const char*>(&target_chars[target_off]);
+
+                if (std::string_view(src_raw_v, src_len) == std::string_view(target_raw_v, target_len)) {
+                    ++count;
+                } else {
+                    if (src_len == 1) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_remove.h:206]: no need to process empty string

File: be/src/vec/functions/array/function_array_remove.h (hunk 10 above)
Comment (xy720) [reply be/src/vec/functions/array/function_array_remove.h:124]: done

**Comment by cambyzju** [comment]
LGTM

**Comment by yangzhg** [comment]
check the p0 reg and BE Code Formatter /

**Review by yangzhg — DISMISSED** [review]
LGTM

**Review by yangzhg — APPROVED** [review]
LGTM

**Pull request merged**